source/server/filter_chain_manager_impl.cc

```diff
@@ -215,6 +215,158 @@ void FilterChainManagerImpl::addFilterChain(
+  for (const auto& filter_chain : filter_chain_span) {
+    const auto& filter_chain_match = filter_chain->filter_chain_match();
+    if (!filter_chain_match.address_suffix().empty() || filter_chain_match.has_suffix_len()) {
+      throw EnvoyException(fmt::format("error adding listener '{}': contains filter chains with "
+                                       "unimplemented fields",
+                                       address_->asString()));
+    }
+    if (filter_chains.find(filter_chain_match) != filter_chains.end()) {
+      throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
+                                       "the same matching rules are defined",
+                                       address_->asString()));
+    }
+    filter_chains.insert(filter_chain_match);
+
+    // Validate IP addresses.
+    std::vector<std::string> destination_ips;
+    destination_ips.reserve(filter_chain_match.prefix_ranges().size());
+    for (const auto& destination_ip : filter_chain_match.prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(destination_ip);
+      destination_ips.push_back(cidr_range.asString());
+    }
+
+    std::vector<std::string> source_ips;
+    source_ips.reserve(filter_chain_match.source_prefix_ranges().size());
+    for (const auto& source_ip : filter_chain_match.source_prefix_ranges()) {
+      const auto& cidr_range = Network::Address::CidrRange::create(source_ip);
+      source_ips.push_back(cidr_range.asString());
+    }
+
+    // Reject partial wildcards, we don't match on them.
+    for (const auto& server_name : filter_chain_match.server_names()) {
+      if (server_name.find('*') != std::string::npos &&
+          !FilterChainManagerImpl::isWildcardServerName(server_name)) {
+        throw EnvoyException(
+            fmt::format("error adding listener '{}': partial wildcards are not supported in "
+                        "\"server_names\"",
+                        address_->asString()));
+      }
+    }
+
+    // Reuse created filter chain if possible.
+    // FilterChainManager maintains the lifetime of FilterChainFactoryContext
+    // ListenerImpl maintains the dependencies of FilterChainFactoryContext
+    auto filter_chain_impl = findExistingFilterChain(*filter_chain);
+    if (filter_chain_impl == nullptr) {
+      // if the build option for this filter chain is build_on_demand, we build a fake placeholder
+      // for now. Otherwise, we build the filter chain directly.
+      // TODO(ASOPVII): add build option to filter chain config.
+      bool build_on_demand = true; // filter_chain.build_option()
+      if (build_on_demand) {
+        filter_chain_impl =
+            std::make_unique<FilterChainImpl>(filter_chain); // fake filter chain placeholder
+      } else {
+        filter_chain_impl =
+            filter_chain_factory_builder.buildFilterChain(*filter_chain, context_creator);
+      }
+      ++new_filter_chain_size;
+    }
```

Comment:
During Ctor of ```listenerImpl```, ```buildFilterChain()``` calls ```filter_chain_manager_.addFakeFilterChains()```. Then build filter chain according to the option ```<on demand or not>```. If ```on demand```, we first build an empty placeholder. Insert this filter chain place holder into the match trie.